[CSS] visited color should not be computed only for links

mdubet opened the pull request:
#### f93d71021b6613c26c7db8da6be03f88169d32c5
<pre>
[CSS] visited color should not be computed only for links
<a href="https://bugs.webkit.org/show_bug.cgi?id=264109">https://bugs.webkit.org/show_bug.cgi?id=264109</a>
<a href="https://rdar.apple.com/115289075">rdar://115289075</a>

Reviewed by Antti Koivisto and Tim Nguyen.

Before this patch, we are computing the visited link color only
when inside a link which creates a bug when we should inherit a computed color from another kind of element.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/currentcolor-visited-fallback-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/currentcolor-visited-fallback-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/currentcolor-visited-fallback.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/color-inherit-link-visited-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/color-inherit-link-visited-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/color-inherit-link-visited.html: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-elem-41-t-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-elem-78-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/animate-elem-41-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/animate-elem-78-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-elem-41-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-elem-78-t-manual-expected.txt:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::applyPropertyToVisitedLinkStyle const):

Canonical link: <a href="https://commits.webkit.org/270270@main">https://commits.webkit.org/270270@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e4c926e7e23ea3e4beb906a9d42b6c7834be06b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25026 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3567 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26279 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27142 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22975 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25294 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5269 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1007 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23249 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25270 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/2605 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21602 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27722 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2300 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22538 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28663 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22835 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22894 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26489 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2243 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/545 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3525 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/22288 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5994 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2688 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2586 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
r=me assuming perf A / B test is OK with this

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2431,6 +2431,7 @@ const Color& RenderStyle::visitedLinkColor() const
 {
+    SET_VAR(m_inheritedData, visitedLinkColor, v);
     SET_VAR(m_inheritedData, color, v);
 }
```
This doesn't seem correct. Visited and unvisited colors are resolved separately. Setting visited link color is handled by BuilderCustom::applyValueColor which has special handling for `currentcolor`, treating it the same as `inherit`. There is a discrepency here:

```        
if (builderState.applyPropertyToVisitedLinkStyle()) {
    auto color = builderState.colorFromPrimitiveValue(primitiveValue, ForVisitedLink::Yes);
    builderState.style().setVisitedLinkColor(color.resolveColor(builderState.parentStyle().visitedLinkColor()));
}
```
vs.
```
    if (builderState.applyPropertyToVisitedLinkStyle())
        builderState.style().setVisitedLinkColor(builderState.parentStyle().color());
```
(inheriting from visitedLinkColor vs color)

anttijk replied:
I suppose the underlying issue is that we don't compute visited link color unless we are inside link

```
inline void Builder::applyCascadeProperty(const PropertyCascade::Property& property)
{
    SetForScope levelScope(m_state.m_currentProperty, &property);

    auto applyWithLinkMatch = [&](SelectorChecker::LinkMatchMask linkMatch) {
        if (property.cssValue[linkMatch]) {
            SetForScope scopedLinkMatchMutation(m_state.m_linkMatch, linkMatch);
            applyProperty(property.id, *property.cssValue[linkMatch], linkMatch);
        }
    };

    applyWithLinkMatch(SelectorChecker::MatchDefault);

    if (m_state.style().insideLink() == InsideLink::NotInside)
        return;

    applyWithLinkMatch(SelectorChecker::MatchLink);
    applyWithLinkMatch(SelectorChecker::MatchVisited);

    m_state.m_linkMatch = SelectorChecker::MatchDefault;
}
```

but also don't know we haven't computed it so we just end up inheriting the non-computed color.

anttijk replied:
Would something like

```
diff --git a/Source/WebCore/style/StyleBuilder.cpp b/Source/WebCore/style/StyleBuilder.cpp
index c725151eb7fc..f617496e0914 100644
--- a/Source/WebCore/style/StyleBuilder.cpp
+++ b/Source/WebCore/style/StyleBuilder.cpp
@@ -344,7 +344,7 @@ void Builder::applyProperty(CSSPropertyID id, CSSValue& value, SelectorChecker::
 
     ASSERT(!isInherit || !isInitial); // isInherit -> !isInitial && isInitial -> !isInherit
 
-    if (m_state.applyPropertyToVisitedLinkStyle() && !isValidVisitedLinkProperty(id)) {
+    if (!m_state.applyPropertyToRegularStyle() && !isValidVisitedLinkProperty(id)) {
         // Limit the properties that can be applied to only the ones honored by :visited.
         return;
     }
diff --git a/Source/WebCore/style/StyleBuilderState.h b/Source/WebCore/style/StyleBuilderState.h
index 891cd9c6c20c..0421fb33914a 100644
--- a/Source/WebCore/style/StyleBuilderState.h
+++ b/Source/WebCore/style/StyleBuilderState.h
@@ -85,7 +85,7 @@ public:
 
     // FIXME: These are mutually exclusive, clean up the code to take that into account.
     bool applyPropertyToRegularStyle() const { return m_linkMatch != SelectorChecker::MatchVisited; }
-    bool applyPropertyToVisitedLinkStyle() const { return m_linkMatch == SelectorChecker::MatchVisited; }
+    bool applyPropertyToVisitedLinkStyle() const { return m_linkMatch != SelectorChecker::MatchLink; }
 
     bool useSVGZoomRules() const;
     bool useSVGZoomRulesForLength() const;
```

work?

mdubet replied:
Yes it's way cleaner. 👍  I'm running AB testing perf and memory because we are gonna compute the property value way more often now.

anttijk commented on LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-elem-41-t-manual-expected.txt:
```diff
@@ -100,7 +100,7 @@ layer at (0,0) size 800x600
               RenderSVGPath {polyline} at (455,150) size 38x36 [stroke={[type=SOLID] [color=#CC0066]}] [fill={[type=SOLID] [color=#CC0066]}] [points="20 10 0 10 15 20 10 2 5 20 20 10"]
             RenderSVGContainer {a} at (638,150) size 38x36 [transform={m=((1.00,0.00)(0.00,1.00)) t=(220.00,0.00)}]
-              RenderSVGPath {polyline} at (638,150) size 38x36 [stroke={[type=SOLID] [color=#CC0066]}] [fill={[type=SOLID] [color=#000000]}] [points="20 10 0 10 15 20 10 2 5 20 20 10"]
+              RenderSVGPath {polyline} at (638,150) size 38x36 [stroke={[type=SOLID] [color=#CC0066]}] [fill={[type=SOLID] [color=#CC0066]}] [points="20 10 0 10 15 20 10 2 5 20 20 10"]
```
Did you confirm these are progressions?

mdubet replied:
Yes, they are fixing the same bug indeed. 

```
<g id="setEleven" color="rgb(204,0,102)" transform="translate(0,200)">
...
     <a xlink:href="" transform="translate(220, 0)" fill="currentColor">
```

nt1m replied:
The testcase in question isn't using currentColor, but it is a similar bug with applying a `visitedLinkFillPaintColor` that is uninitialized. Here's a reduced testcase:
```
<svg width="100" height="100" fill="red" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
  <g fill="green">
    <a xlink:href="">
      <rect width="100" height="100"></rect>
    </a>
  </g>
</svg>
```

We should add a WPT for this.

anttijk commented on Source/WebCore/style/StyleBuilderState.h:
```diff
@@ -83,9 +83,8 @@ class BuilderState {
-    // FIXME: These are mutually exclusive, clean up the code to take that into account.
     bool applyPropertyToRegularStyle() const { return m_linkMatch != SelectorChecker::MatchVisited; }
-    bool applyPropertyToVisitedLinkStyle() const { return m_linkMatch == SelectorChecker::MatchVisited; }
+    bool applyPropertyToVisitedLinkStyle() const { return m_linkMatch != SelectorChecker::MatchLink; }
```
We should also (separately) fix all the applyInheritFoo functions with visited link handling (generated and otherwise) to inherit from the visited link style.

mdubet replied:
https://bugs.webkit.org/show_bug.cgi?id=264252

mdubet committed:
[CSS] visited color should not be computed only for links
https://bugs.webkit.org/show_bug.cgi?id=264109
rdar://115289075

Reviewed by Antti Koivisto and Tim Nguyen.

Before this patch, we are computing the visited link color only
when inside a link which creates a bug when we should inherit a computed color from another kind of element.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/currentcolor-visited-fallback-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/currentcolor-visited-fallback-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/currentcolor-visited-fallback.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/color-inherit-link-visited-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/color-inherit-link-visited-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/color-inherit-link-visited.html: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-elem-41-t-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/animate-elem-78-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/animate-elem-41-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/animate-elem-78-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-elem-41-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/animate-elem-78-t-manual-expected.txt:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::applyPropertyToVisitedLinkStyle const):

Canonical link: https://commits.webkit.org/270270@main
